pkg/profiles/tap.go

```diff
@@ -0,0 +1,142 @@
+package profiles
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/linkerd/linkerd2/controller/api/util"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// RenderTapOutputProfile performs a tap on the desired resource and generates
+// a service profile with routes pre-populated from the tap data
+// Only inbound tap traffic is considered.
+func RenderTapOutputProfile(client pb.ApiClient, tapResource, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int, w io.Writer) error {
+	requestParams := util.TapRequestParams{
+		Resource:  tapResource,
+		Namespace: namespace,
+	}
+	log.Debugf("Running `linkerd tap %s --namespace %s`", tapResource, namespace)
+
+	req, err := util.BuildTapByResourceRequest(requestParams)
+	if err != nil {
+		return err
+	}
+
+	profile, err := tapToServiceProfile(client, req, namespace, name, controlPlaneNamespace, tapDuration, routeLimit)
+	if err != nil {
+		return err
+	}
+
+	output, err := yaml.Marshal(profile)
+	if err != nil {
+		return fmt.Errorf("Error writing Service Profile: %s", err)
+	}
+	w.Write(output)
+	return nil
+}
+
+func tapToServiceProfile(client pb.ApiClient, tapReq *pb.TapByResourceRequest, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int) (sp.ServiceProfile, error) {
+	profile := sp.ServiceProfile{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace),
+			Namespace: controlPlaneNamespace,
+		},
+		TypeMeta: meta_v1.TypeMeta{
+			APIVersion: "linkerd.io/v1alpha1",
+			Kind:       "ServiceProfile",
+		},
+	}
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(tapDuration))
+	defer cancel()
+
+	tapClient, err := client.TapByResource(context.Background(), tapReq)
+	if err != nil {
+		return profile, err
+	}
+
+	routes := routeSpecFromTap(ctx, tapClient, routeLimit)
+
+	profile.Spec.Routes = routes
+
+	return profile, nil
+}
+
+func routeSpecFromTap(ctx context.Context, tapClient pb.Api_TapByResourceClient, routeLimit int) []*sp.RouteSpec {
+	routes := make([]*sp.RouteSpec, 0)
+	routesMap := make(map[string]*sp.RouteSpec)
+
+recvLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			break recvLoop
+		default:
+			log.Debug("Waiting for data...")
+			event, err := tapClient.Recv()
+
+			if err == io.EOF {
+				break recvLoop
+			}
+			if err != nil {
+				fmt.Fprintln(os.Stderr, err)
```

Comment:
I think the reason this is here is so that it prints out the error even if you don't have a log level set? if I do this instead, it fails silently without printing the error (the other cli/ commands use this pattern as well)

Reply:
Out of scope for this issue, but maybe we should investigate this separately.  It seems weird that there isn't a default log level??

Reply:
logrus's default is `InfoLevel`. in the cli, we expose a `--verbose` flag, when disabled sets `PanicLevel`, when enabled, sets `DebugLevel`.

i agree with @rmars implementation here, that this error should always print regardless of log level, and should not print like a log message, but rather a cli output message.